task_processing/plugins/kubernetes/kube_client.py

```diff
@@ -14,12 +23,133 @@ def __init__(self, kubeconfig_path: Optional[str] = None) -> None:
                 "or pass a value for `kubeconfig_path`!"
             )
 
+        self.kubeconfig_path = kubeconfig_path
+        self.kubecontext = os.environ.get("KUBECONTEXT")
+
         kube_config.load_kube_config(
-            config_file=kubeconfig_path,
-            context=os.environ.get("KUBECONTEXT")
+            config_file=self.kubeconfig_path,
+            context=self.kubecontext
         )
 
         # any Kubernetes APIs that we use should be added as members here (much like as we
         # do in the KubeClient class in PaaSTA) to ensure that they're only used after we've
-        # loaded the relevant kubeconfig
+        # loaded the relevant kubeconfig. Additionally, they should also be recreated in
+        # reload_kubeconfig() so that changes in certs are picked up.
+        self.core = kube_client.CoreV1Api()
+
+    def reload_kubeconfig(self) -> None:
+        """
+        Helper meant to be called after a 401 from apiserver.
+
+        Since we rotate certs regularly internally and the kubernetes clientlib doesn't
+        automatically watch for changes to the certs specified by the kubeconfig file, we
+        need to manually reload said file as well as recreate any API classes (as otherwise
+        they don't pick up the new configuration)
+        """
+        kube_config.load_kube_config(
+            config_file=self.kubeconfig_path,
+            context=self.kubecontext
+        )
         self.core = kube_client.CoreV1Api()
+
+    def maybe_reload_on_exception(self, exception: Exception) -> bool:
+        """
+        Small wrapper around KubeClient::reload_kubeconfig() to centralize when to reload kubeconfig
+        on exceptions.
+
+        Currently, the only exception that will cause a reload is an ApiException with an
+        HTTP 401 (unauthorized) status.
+        """
+        if isinstance(exception, ApiException):
+            if exception.status == HTTPStatus.UNAUTHORIZED.value:
+                logger.info(
+                    "Recieved UNAUTHORIZED response from apiserver - assuming certs have "
+                    "expired and reloading."
+                )
+                self.reload_kubeconfig()
```

Comment:
Maybe we can have some kind of confirmation maybe that reloading the kubeconfig file worked successfully? Load_kube_config doesn't return anything, but what happens if load_kube_config file doesn't find the file for example. Not sure if this is overkill though.

Reply:
in that case, the k8s clientlib would throw an exception - i'm not quite sure what we want to do in that case though - maybe once we have event stuff working we can catch that exception and send a control event in that case and tell tron to shut us down?

Reply:
Yup. I think that sounds reasonable 👍